### DISCUSSION TITLE
HALMET example firmware now available

### DISCUSSION THREAD

**Discussion by mairas** [hatlabs/discussions#83] [Show and tell]
Hi all,

I just finished the initial example firmware for [HALMET](https://shop.hatlabs.fi/products/halmet). The firmware gives an example how to read both the analog and the digital inputs. For digital inputs, both RPM and alarm inputs are provided.

The data is output both to Signal K and to NMEA 2000 engine parameter and fluid level messages.

I would love to hear your comments, suggestions and improvement ideas for the code. See here:

https://github.com/hatlabs/HALMET-example-firmware

**Comment by Stonepaiste** [comment]
Awesome! When I try to build without any edit  I get a problem with "no matching function for call to 'ArduinoJson::V700PB2::JsonObject::createNestedObject(String&) const'"

I know this is an example firmware. But I'm a little unsure if it should work out of the box with this firmware and the functions in it or if it needs more tinkering before build just to test it. 

My plan is to use it as the example code is written but add tempsensor functionality as well. 
Best regards. 
Mathias

**Reply by mairas** [reply]
It's supposed to work as-is. I'll have a look tomorrow!

**Comment by Stonepaiste** [comment]
I think I found the problem. When I opened the example project in visual studio code platform.io<http://platform.io> downloaded the json 7.0 instead of 6.21 and since they made a lot of changes in 7.0 it broke the code. I guess. So I downgraded to 6.21 in stead and it seems to build now.

On 6 Jan 2024, at 19.18, Matti Airas ***@***.***> wrote:



It's supposed to work as-is. I'll have a look tomorrow!

**Reply by mairas** [reply]
Yes, that was it! A quick fix was to pin the version in SensESP to 6.x (which is a sensible thing to do, anyways) and release a new version. 

I also updated the HALMET example code - it should be a bit cleaner now.

**Comment by Stonepaiste** [comment]
Sounds good will take a look.
Gotta say awesome product! It’s really something iv’e been waiting for so thank you for that!

A quick question - Which might not be that quick. If I want to add tempsensors to the 1-wire port. Can I then take the code from your ESP 32 tempsenor project or would that not work. And would I need the whole code or just parts of it. The pins I need to solder in do you have some of those, couldn’t find it in your store. ? Or is that something I should be able to find anywhere . I just put in a new order for a NMEA2000 male pin and some other parts today, so if you had it you might be able to put that in the box. (I’ll of course pay).

-Mathias

On 7 Jan 2024, at 17.47, Matti Airas ***@***.***> wrote:



Yes, that was it! A quick fix was to pin the version in SensESP to 6.x (which is a sensible thing to do, anyways) and release a new version.

I also updated the HALMET example code - it should be a bit cleaner now.

**Reply by mairas** [reply]
Take just the parts you need. What you need depends on whether you're interested in Signal K, NMEA 2000, or both.

In any case:

- Add the dependency to platformio.ini
- Add the sensesp_onewire include in main.cpp
- Add the onewire input code
- Add the outputs you want to use
- Connect the inputs to the outputs outputs
- You might want to leave the display code out (or merge them together)

About the header connectors: the idea was that the HALMET sales package included the headers, but I forgot to add them to the BOM at first, so my assistant didn't know to add them. My mistake. I'll make sure you'll get the headers together with your new order.

**Comment by Stonepaiste** [comment]
Thank you for the answer. It’s both NMEA and Signal k I’m after.

I made a build today which compiles and when I connect the Hamlet I can see all the temperature sensor identifiers (whithout any sensors connected to the board ) for a while then they disappears.
When you write Outputs I want to use and connect the inputs to the outputs. Is that within the main or in some of the dependencies ?

No worries about the headers. Looking forward to to getting the rest so I hopefully can figure it out.

Best regards


On 7 Jan 2024, at 21.59, Matti Airas ***@***.***> wrote:



Take just the parts you need. What you need depends on whether you're interested in Signal K, NMEA 2000, or both.

In any case:

  *   Add the dependency to platformio.ini
  *   Add the sensesp_onewire include in main.cpp
  *   Add the onewire input code
  *   Add the outputs you want to use
  *   Connect the inputs to the outputs outputs
  *   You might want to leave the display code out (or merge them together)

About the header connectors: the idea was that the HALMET sales package included the headers, but I forgot to add them to the BOM at first, so my assistant didn't know to add them. My mistake. I'll make sure you'll get the headers together with your new order.

**Comment by addboy58** [comment]
Is there a go by for analog engine sensors? I would like to monitor, RPMs, Voltage, water temp and oil pressure.

**Comment by big-col** [comment]
Hi, I'm getting the following error when i try and upload the example firmware (unedited) .  is this something ive done wrong?

#include errors detected. Please update your includePath. Squiggles are disabled for this translation unit (C:\Users\colin\OneDrive\Boat\Halmet\HALMET-example-firmware-main\src\main.cpp).C/C++(1696)
cannot open source file "esp_mac.h" (dependency of "halmet_serial.h")C/C++(1696)

**Reply by mairas** [reply]
This is not immediately obvious to me. It sounds like PlatformIO has failed to install the Espressif32 (ESP32) platform files. But if that was the case, it would refuse to build at all.

As a first aid, try removing the `.pio` directory under your project directory (`C:\Users\colin\OneDrive\Boat\Halmet\HALMET-example-firmware-main`). That directory contains all project-specific libraries, dependencies and build files. Additionally, remove `.platformio`, probably under `C:\Users\colin` (at least on Mac and Linux it's under the user home directory, not 100% sure about Windows). That contains all platform SDKs and tools. Both are automatically re-created when you open and build the project in VSCode.

**Reply by big-col** [reply]
Thanks I'll take a look. In the meantime, do you think this post is related?  https://github.com/microsoft/vscode-arduino/issues/1687

**Reply by mairas** [reply]
No, intellisense is a different thing.

**Reply by big-col** [reply]
Fixed!  deleting as you requested fixed the problem :)  thanks for a speedy response
